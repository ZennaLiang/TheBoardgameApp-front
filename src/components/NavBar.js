import React from "react";
import { NavLink, withRouter } from "react-router-dom"; // withRouter to access history location-URL link
import { signout, isAuthenticated } from "../auth";
import BgLogo from "../images/BgLogo.png";
//import Notification from "./notifications/Notification";
import { getEventsByUserId } from "../calendar/apiCalendar";
import { getAllTradeRequestsById } from "../trades/apiTrade";

class NavBar extends React.Component {
  // constructor() {
  //   super();
  //   // if (isAuthenticated().user) {
  //   //   this.state = {
  //   //     notifications: this.getNotifications()
  //   //   };
  //   // } else {
  //   //   this.state = {
  //   //     notifications: []
  //   //   };

  //   // }
  // }

  getNotifications = async () => {
    var notifications = [];
    await getEventsByUserId(isAuthenticated().user._id, isAuthenticated().token)
      .then((event) => {
        event.map((e) => {
          notifications.push({
            id: e._id,
            name: e.title,
            type: "Event",
            link: "/calendar/" + isAuthenticated().user._id,
            isRead: false,
          });
        });
      })
      .then(
        await getAllTradeRequestsById(isAuthenticated().user._id).then(
          (trade) => {
            trade.map((t) => {
              notifications.push({
                id: t._id,
                name: t.tradeReceiver.name,
                type: t.status.concat(" Trade"),
                link: "/trades",
                isRead: false,
              });
            });
            console.log(notifications);
          }
        )
      );
    return notifications;
  };

  render() {
    return (
      <nav className="navbar navbar-icon-top navbar-expand-lg navbar-light bg-light sticky-top shadow-sm">
        <div className="container-fluid">
          <NavLink
            className="navbar-brand"
            to={isAuthenticated() ? "/posts" : "/"}
          >
            <img src={BgLogo} width="20" height="20" alt="" /> Boardgame Guru
          </NavLink>
          <button
            className="navbar-toggler"
            type="button"
            data-toggle="collapse"
            data-target="#bgNavBar"
            aria-controls="bgNavBar"
            aria-expanded="false"
            aria-label="Toggle navigation"
          >
            <span className="navbar-toggler-icon"></span>
          </button>
          <div className="collapse navbar-collapse" id="bgNavBar">
            {isAuthenticated() && (
              <ul className="navbar-nav mr-auto">
                <li className="nav-item ">
                  <NavLink
                    className="nav-link"
                    activeClassName="selected"
                    to={`/user/${isAuthenticated().user._id}`}
                  >{`${isAuthenticated().user.name}'s Profile`}</NavLink>
                </li>
                <li className="nav-item ">
                  <NavLink
                    className="nav-link"
                    activeClassName="selected"
                    to="/posts"
                  >
                    Posts
                  </NavLink>
                </li>
                {/* <li className="nav-item ">
                  <NavLink
                    className="nav-link"
                    activeClassName="selected"
                    to="/users"
                  >
                    Users
                  </NavLink>
                </li> */}
                <li className="nav-item dropdown ">
                  <a
                    className="nav-link dropdown-toggle"
                    href="/#"
                    id="navbarDropdownCollectionLink"
                    role="button"
                    data-toggle="dropdown"
                    aria-haspopup="true"
                    aria-expanded="false"
                  >
                    Collection
                  </a>
                  <div
                    className="dropdown-menu"
                    aria-labelledby="navbarDropdownCollectionLink"
                  >
                    <NavLink
                      className="dropdown-item"
                      activeClassName="selected"
                      to="/collection/bgguru"
                    >
                      Guru Collection
                    </NavLink>
                    <NavLink
                      className="dropdown-item"
                      activeClassName="selected"
                      to="/collection/bgg"
                    >
                      BGGeek Collection
                    </NavLink>
                  </div>
                </li>
                <li className="nav-item ">
                  <NavLink
                    className="nav-link"
                    activeClassName="selected"
                    to="/trades"
                  >
                    Trades
                  </NavLink>
                </li>
                <li className="nav-item ">
                  {isAuthenticated().user && (
                    <NavLink
                      className="nav-link"
                      activeClassName="selected"
                      to={`/calendar/${isAuthenticated().user._id}`}
                    >
                      Calendar
                    </NavLink>
                  )}
                </li>
              </ul>
            )}
            {/* {isAuthenticated().user && this.state.notifications && (
              <Notification notificationsObj={this.state.notifications} />
            )} */}
            <ul className="navbar-nav ">
              {!isAuthenticated() && (
                <>
                  <li className="nav-item">
                    <NavLink
                      className="nav-link"
                      activeClassName="selected"
                      to="/signin"
                    >
                      Sign In
                    </NavLink>
                  </li>
                  <li className="nav-item ">
                    <NavLink
                      className="nav-link"
                      activeClassName="selected"
                      to="/signup"
                    >
                      Sign Up
                    </NavLink>
                  </li>
                </>
              )}
              {isAuthenticated() && (
                <>
                  <li className="nav-item dropdown">
                    <a
                      className="nav-link"
                      href="/#"
                      id="navbarDropdown"
                      role="button"
                      data-toggle="dropdown"
                      aria-haspopup="true"
                      aria-expanded="false"
                    >
                      <div
                        className=" profileInitalCircle text-center"
                        style={{ fontSize: "32px" }}
                      >
                        <strong>{`${isAuthenticated().user.name.substring(
                          0,
                          1
                        )}`}</strong>
                      </div>
                    </a>
                    <div
                      className="dropdown-menu"
                      aria-labelledby="navbarDropdown"
                    >
                      {isAuthenticated().user && (
                        <NavLink
                          className="dropdown-item"
                          activeClassName="selected"
                          to={`/user/edit/${isAuthenticated().user._id}`}
                        >
                          Settings
                        </NavLink>
                      )}
                      {isAuthenticated() &&
                        isAuthenticated().user.role === "admin" && (
                          <NavLink
                            className="dropdown-item"
                            activeClassName="selected"
                            to="/admin"
                          >
                            Admin
                          </NavLink>
                        )}
                      <div className="dropdown-divider"></div>
                      <span
                        className="dropdown-item"
                        data-toggle="tooltip"
                        title="Sign Out"
                        aria-label="Sign Out"
                        onClick={() =>
                          signout(() => window.location = "/")
                        }
                        style={{ cursor: "pointer" }}
                      >
                        Sign Out
                      </span>
                    </div>
                  </li>
                </>
              )}
            </ul>
          </div>
        </div>
      </nav>
    );
  }
}

export default withRouter(NavBar);
