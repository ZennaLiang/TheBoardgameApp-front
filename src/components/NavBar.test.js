import React from "react";
import NavBar from "./NavBar";
import { BrowserRouter, NavLink } from "react-router-dom";
import { shallow, mount } from "enzyme";
import { render, screen } from "@testing-library/dom";
import "../../setupTests";

const links = [
  { text: "Sign In", location: "/signin" },
  { text: "Sign Up", location: "/signup" }
];
const tree = (
  <BrowserRouter>
    <NavBar />
  </BrowserRouter>
);
test.each(links)("Check if Nav Bar have %s link.", link => {
  //Ensure the text is in the dom, will throw error it can't find

  const linkDom = <NavLink to={link.location}>{link.text}</NavLink>;
  const wrapper = shallow(tree);
  expect(wrapper.find(linkDom)).toBeTruthy();
});

test("Check if NavBar matches snapshot", () => {
  const wrapper = mount(tree);
  expect(wrapper).toMatchSnapshot();
});
