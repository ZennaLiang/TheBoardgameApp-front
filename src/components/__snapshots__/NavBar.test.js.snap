// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Check if NavBar matches snapshot 1`] = `
<BrowserRouter>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "createHref": [Function],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <withRouter(NavBar)>
      <NavBar
        history={
          Object {
            "action": "POP",
            "block": [Function],
            "createHref": [Function],
            "go": [Function],
            "goBack": [Function],
            "goForward": [Function],
            "length": 1,
            "listen": [Function],
            "location": Object {
              "hash": "",
              "pathname": "/",
              "search": "",
              "state": undefined,
            },
            "push": [Function],
            "replace": [Function],
          }
        }
        location={
          Object {
            "hash": "",
            "pathname": "/",
            "search": "",
            "state": undefined,
          }
        }
        match={
          Object {
            "isExact": true,
            "params": Object {},
            "path": "/",
            "url": "/",
          }
        }
      >
        <div>
          <nav
            className="navbar navbar-icon-top navbar-expand-lg navbar-light bg-light sticky-top shadow-sm"
          >
            <div
              className="container-fluid"
            >
              <NavLink
                className="navbar-brand"
                data-testid="company-logo"
                to="/"
              >
                <Link
                  aria-current="page"
                  className="navbar-brand active"
                  data-testid="company-logo"
                  style={Object {}}
                  to={
                    Object {
                      "hash": "",
                      "pathname": "/",
                      "search": "",
                      "state": null,
                    }
                  }
                >
                  <LinkAnchor
                    aria-current="page"
                    className="navbar-brand active"
                    data-testid="company-logo"
                    href="/"
                    navigate={[Function]}
                    style={Object {}}
                  >
                    <a
                      aria-current="page"
                      className="navbar-brand active"
                      data-testid="company-logo"
                      href="/"
                      onClick={[Function]}
                      style={Object {}}
                    >
                      <img
                        alt="Company Logo"
                        height="20"
                        src={
                          Object {
                            "process": [Function],
                          }
                        }
                        width="20"
                      />
                       
                      Boardgame Guru
                    </a>
                  </LinkAnchor>
                </Link>
              </NavLink>
              <button
                aria-controls="bgNavBar"
                aria-expanded="false"
                aria-label="Toggle navigation"
                className="navbar-toggler"
                data-target="#bgNavBar"
                data-toggle="collapse"
                type="button"
              >
                <span
                  className="navbar-toggler-icon"
                />
              </button>
              <div
                className="collapse navbar-collapse"
                id="bgNavBar"
              >
                <ul
                  className="navbar-nav "
                >
                  <li
                    className="nav-item"
                  >
                    <NavLink
                      activeClassName="selected"
                      className="nav-link"
                      to="/signin"
                    >
                      <Link
                        aria-current={null}
                        className="nav-link"
                        to={
                          Object {
                            "hash": "",
                            "pathname": "/signin",
                            "search": "",
                            "state": null,
                          }
                        }
                      >
                        <LinkAnchor
                          aria-current={null}
                          className="nav-link"
                          href="/signin"
                          navigate={[Function]}
                        >
                          <a
                            aria-current={null}
                            className="nav-link"
                            href="/signin"
                            onClick={[Function]}
                          >
                            Sign In
                          </a>
                        </LinkAnchor>
                      </Link>
                    </NavLink>
                  </li>
                  <li
                    className="nav-item "
                  >
                    <NavLink
                      activeClassName="selected"
                      className="nav-link"
                      to="/signup"
                    >
                      <Link
                        aria-current={null}
                        className="nav-link"
                        to={
                          Object {
                            "hash": "",
                            "pathname": "/signup",
                            "search": "",
                            "state": null,
                          }
                        }
                      >
                        <LinkAnchor
                          aria-current={null}
                          className="nav-link"
                          href="/signup"
                          navigate={[Function]}
                        >
                          <a
                            aria-current={null}
                            className="nav-link"
                            href="/signup"
                            onClick={[Function]}
                          >
                            Sign Up
                          </a>
                        </LinkAnchor>
                      </Link>
                    </NavLink>
                  </li>
                </ul>
              </div>
            </div>
          </nav>
          <GettingstartedAlert />
        </div>
      </NavBar>
    </withRouter(NavBar)>
  </Router>
</BrowserRouter>
`;
